src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1050,132 +1071,667 @@ private StringConcatFactory() {
+                        int paramSlot = paramSlots[i];
+                        var kind = TypeKind.from(cl);
+                        if (needStringOf(cl)) {
+                            paramSlot = stringSlots[i];
+                            kind = TypeKind.from(String.class);
+                        }
+                        cb.loadLocal(kind, paramSlot);
+                    }
+                    cb.invokestatic(concatClass, "prepend", ConstantUtils.methodTypeDesc(prependArgs));
+
+                    // return new String(buf, coder);
+                    cb.new_(CD_String)
+                      .dup()
+                      .aload(bufSlot)
+                      .iload(coderSlot)
+                      .invokespecial(CD_String, INIT_NAME, MTD_STRING_INIT)
+                      .areturn();
+                }
+            };
+        }
+
+        /**
+         * Generate coder method method. <p>
+         *
+         * The following is an example of the generated target code:
+         *
+         * <blockquote><pre>
+         * import static java.lang.StringConcatHelper.stringSize;
+         *
+         * static int length(int length, int arg0, long arg1, boolean arg2, char arg3,
+         *                  String arg4, String arg5, String arg6, String arg7) {
+         *     return length + stringSize(arg0) + stringSize(arg1) + stringSize(arg2) + 1 + stringSize(arg4)
+         *               + stringSize(arg5) + stringSize(arg6) + stringSize(arg7);
+         * }
+         * </pre></blockquote>
+         */
+        private static Consumer<CodeBuilder> generateLengthMethod(MethodType lengthArgs) {
+            return new Consumer<CodeBuilder>() {
+                @Override
+                public void accept(CodeBuilder cb) {
+                    int paramCount = lengthArgs.parameterCount(),
+                        lengthSlot = 0,
+                        nextSlot   = 1;
+                    cb.iload(lengthSlot);
+                    for (int i = 1; i < paramCount; i++) {
+                        var cl   = lengthArgs.parameterType(i);
+                        var kind = TypeKind.from(cl);
+                        MethodTypeDesc methodTypeDesc;
+                        if (cl == char.class) {
+                            cb.iconst_1();
+                        } else {
+                            if (cl == int.class) {
+                                methodTypeDesc = MTD_int_int;
+                            } else if (cl == long.class) {
+                                methodTypeDesc = MTD_int_long;
+                            } else if (cl == boolean.class) {
+                                methodTypeDesc = MTD_int_boolean;
+                            } else {
+                                methodTypeDesc = MTD_int_String;
+                            }
+                            cb.loadLocal(kind, nextSlot)
+                              .invokestatic(CD_StringConcatHelper, "stringSize", methodTypeDesc);
```

Comment:
The approach here isn't checking for overflow at all, deferring to the length check in the newArray step. While current tests all pass with this I think we need to maintain checking for overflow as we go, otherwise it'd be possible to construct concatenations where the length overflows to negative and then back again to positive. Let's try this with a new set of `mix` methods.